File: Interleaving Positive and Negative Numbers.cpp
class Solution {
public:
    /**
     * @param A: An integer array.
     * @return: void
     */
    void rerange(vector<int> &A) {
        // write your code here
        // sort(A.begin(), A.end());
        // int mid = 0;
        // while (A[mid] < 0)
        // {
        //     ++mid;
        // }
        // if (mid * 2 < A.size())
        // {
        //     reverse(A.begin(), A.end());
        //     mid = A.size() - mid;
        //     swap(A[mid], A[A.size() - 1]);
        // }
        // size_t left = 1;
        // size_t right = 2 * mid - 2;
        // while (left < right)
        // {
        //     swap(A[left], A[right]);
        //     left += 2;
        //     right -= 2;
        // }
        //good solution
        int posCount = 0;
        for (const auto &a : A)
        {
            posCount += a > 0 ? 1 : 0;
        }
        
        int expectPos = 2 * posCount > A.size() ? true : false;
        int pos = 0, neg = 0, i = 0;
        while (i < A.size())
        {
            while (pos < A.size() && A[pos] > 0)
            {
                ++pos;
            }
            while (neg < A.size() && A[neg] < 0)
            {
                ++neg;
            }
            
            if (expectPos && neg < A.size())
            {
                swap(A[i], A[neg]);
                ++i;
                ++neg;
            }
            else if (!expectPos && pos < A.size())
            {
                swap(A[i], A[pos]);
                ++i;
                ++pos;
            }
            expectPos = !expectPos;
        }
    }
};

File: Interleaving Positive and Negative Numbers.py
class Solution:
    """
    @param A: An integer array.
    @return nothing
    """
    def rerange(self, A):
        # write your code here
        
        if not A:
            return None
        length = len(A)
        tag = 1
        ### reduce must be init
        positive = reduce(lambda x, y: x + 1 if y > 0 else x, A, 0)
        if 2 * positive > length:
            tag = -1
        # tranf = lambda x: (1 + 2 * x) % (length | 1)
        # left, right, pivot = 0, length - 1, 0
        
        # while left <= right:
        #     leftIndex = tranf(left)
        #     # print leftIndex
        #     if tag * A[leftIndex] > 0:
        #         pivotIndex = tranf(pivot)
        #         A[pivotIndex], A[leftIndex] = A[leftIndex], A[pivotIndex]
        #         pivot += 1
        #         left += 1
        #     else:
        #         rightIndex = tranf(right)
        #         A[rightIndex], A[leftIndex] = A[leftIndex], A[rightIndex]
        #         right -= 1
        left, pivot, right = 0, 0, length - 1
        while left <= right:
            if tag * A[left] > 0:
                A[pivot], A[left] = A[left], A[pivot]
                pivot += 1
                left += 1
            else:
                A[right], A[left] = A[left], A[right]
                right -= 1
        if length % 2 == 1:
            A[0], A[pivot] = A[pivot], A[0]
            pivot = 2
            mirpivot = length - 2
        else:
            mirpivot = length - 2
            pivot = 1
        while pivot <= mirpivot:
            A[pivot], A[mirpivot] = A[mirpivot], A[pivot]
            pivot += 2
            mirpivot -= 2